proposals/0000-no-toplevel-field-selectors.rst

```diff
@@ -0,0 +1,134 @@
+disambiguation handled by ``DuplicateRecordFields``. The necessary selectors
+will be part of a ``Record(..)`` export, or can also be named one by one. They
+always have to be associated with the data type though, because there is no more
+toplevel selector.
+
+A function for Template Haskell to go from a record field name to a selector for
+that field will be provided, because it's not possible anymore to go from field
+name directly to selector.
+
+Example
+^^^^^^^
+
+Given a data structure
+
+    data Foo = Foo { bar :: Int, baz :: String }
+
+The following will be available:
+
+- the type ``Foo``
+- the constructor ``Foo``
+- the two functions ``bar`` and ``baz``
+- the names ``bar`` and ``baz`` for record construction (``Foo { bar = 3, baz = "foo" }``)
+- the names ``bar`` and ``baz`` for ``RecordWildCards``
+
+If the language extension ``NoToplevelFieldSelectors`` is enabled for the module
+or ``Foo`` specifically, all of the above will be generated, except for the two
+functions ``bar`` and ``baz``.
+
+Wildcard exports will work as before, except for the two functions. Even if
+these functions are otherwise defined, the wildcard will not export them.
+Exporting the names for record construction now has to be specific to the
+record. Without ambiguitiy, previously this was equivalent
+
+    module A where (Foo(Foo, bar, baz))
+
+    module A where (Foo(Foo, bar), baz)
+
+Because of the new semantics, these two export statements are now different.
+Maybe a wildcard export will export functions of the same name as well, this
+question isn't resolved yet.
+
+Effect and Interactions
+-----------------------
+
+Breakage estimation
+^^^^^^^^^^^^^^^^^^^
+
+Enabling this extension will break a lot of Template Haskell. Going from record
+field name to selector won't work anymore. A new way to go from record field
+name to selector has to be found.
+
+Anything that generates code with the help of Generic should be fine. The same
+functionality that generates the anonymous functions for Generic could be used
+to provide TH functionality to replace the existing toplevel functions.
+
+The record extensions NamedFieldPuns, RecordWildCards, DisambiguateRecordFields,
+and DuplicateRecordFields are unaffected by this change.
+
+
+Costs and Drawbacks
+-------------------
+
+This might cause some confusion that record fields can't be accessed by toplevel
+selectors anymore - however, that shouldn't be too big of an issue, because some
+library authors already stopped exporting these selectors so they don't have to
+break downstream software on record changes.
+
+
+Alternatives
+------------
+
+None.
+
+
+Unresolved questions
+--------------------
+
+
+Implementation Plan
+-------------------
+
+For deriving the Generic instances, either ``Record.fieldName``,
+``OverloadedRecordFields`` or an anonymous function which generates the
+corresponding selector in core will be used. (to be clarified)
```

Comment:
I don't understand this sentence. Despite many related happenings, `OverloadedRecordFields` itself is not an extension in GHC HEAD, and I have no idea what `Record.fieldName` refers to.

More generally, I'm happy to advise on implementing this, assuming it gets accepted. My suggestion would be to make this work along the same lines as `DuplicateRecordFields`. This still generates selector functions behind the scenes, but gives them mangled names (to be unique to the module); it then converts between user-visible and mangled names as needed. You'd need to add a flag to fields indicating that they had been defined without the top-level field selectors, then the name resolution code would look at this flag and ignore such fields when resolving names of variables in expressions (but not in other contexts, e.g. record construction). Since the selector function still exists, it can be used by the existing `HasField` stuff, and its name can be used to uniquely identify the field where necessary.

Reply:
So you're suggesting to implement it by generating the functions inside `Generic` by pattern matching? (What's the difference to positional-based access btw?)